### DISCUSSION TITLE
Sticky columns

### DISCUSSION THREAD

**Discussion by muchgitsuchwow** [YaLTeR/niri#149] [Ideas]
Coming from i3/sway I'm used to one window (column) being visible all the time, left on the screen. I use split/tabbed/stacked on the right side. 

In terms of niri this could be a sticky column, e.g. toggled with a bind. 

Feel very free to close this if that doesn't make sense in terms of what niri is/should be or in terms of codebase. Otherwise I'm  looking forward to something like that.

**Comment by YaLTeR** [comment]
Hmm, interesting idea. Stick a column on the side then adjust the struts to shrink the working area for the remaining windows. One thing that's very unclear to me here is how the binds will look like for this kind of setup.

**Comment by muchgitsuchwow** [comment]
Careful: I've totally no idea what I'm talking about.

In my naive assumption there is at least one rule: Only the first column in a workspace may become sticky. If a column was made sticky, it behaves like splitting the screen size niri will manage for scrolling window management. Nothing else changes (for this workspace). Binds should stay the same, e.g. `focus-column-first` should focus the sticky column.

I'm probably missing a LOT of things here. Dunno. Not in a hurry here though. As I said, if it fits, fine. If not, also fine :)

**Comment by superiums** [comment]
as one column may contain many view/window,  maybe sticky view is better than sticky column.
but i have no idea about which is easiear to implemnt.

**Comment by muchgitsuchwow** [comment]
By the way, ... how do I split a column vertically to have multiple windows in one column?

**Reply by YaLTeR** [reply]
You can't right now, there's no splitting columns back into rows.

**Reply by muchgitsuchwow** [reply]
Figured it out. What I meant was `consume-window-into-column` and `expel-window-from-column`. 

@superiums: What's the difference between view and column? I never came across `view`, looking at `config.kdl`, which is my primary source of informationt right now.

**Reply by superiums** [reply]
you can put many window into one column just press `Mod+,`, and switch back via `Mod+.` with default config.
see the screenshot on homepage.

**Comment by Nairou** [comment]
+1, but wanted to add that it would be useful as either the first column in a workspace, or the last column, depending on which side of the screen you wanted to be fixed.

Though in reality the column position doesn't matter, you could select any one column as fixed left or right, and it could move to that position, regardless of whether it was originally the first or last. Whether or not the original column position mattered is an implementation detail.

From there, the rest of the screen scrolls as normal, as if the screen width had been reduced by the sticky window size.

**Comment by victordiaz** [comment]
I was writing another comment and came across this, so I wanted to explain why window pinning is actually pretty useful.

When I'm working on projects that need multiple windows open, I end up scrolling back and forth way too much across my workspace. It makes me lose track of what I was doing and gets pretty confusing.

Picture this: you've got a code editor, browser, documentation, and SQL browser all open. But only 2 windows can fit on your screen at once. The code editor is where you're doing most of your work, but you constantly need to check the browser, look up docs, or check the database browser. If you're switching between all four windows constantly, your main work window keeps disappearing and you lose your place.

If you could "pin" that main window to the left side of your screen, it would stay visible while you work with the other windows. That way you keep your context and don't have to jump around as much.

**This is the normal behavior**

![normal](https://github.com/user-attachments/assets/5ddd75cc-205b-472a-b7ea-4a3ebc2c8607)


**This shows how pinning the first window might be**

![pinned](https://github.com/user-attachments/assets/7be4a304-7779-45c4-a668-1e8be9fe19d4)


Btw, I love niri, its truly an amazing piece of software :)

**Reply by Nairou** [reply]
For that particular use case, tabbed columns would work really well, for the second column.

**Reply by victordiaz** [reply]
Yes, I agree tabs could work!

At the same time tabs use a different stack/modality and makes things a bit confusing to navigate from my point of view.

For example:
- if I am a tab group and I want to add a new window, the window will be added in the normal horizontal stack
- reordering tabs have poorer visual cues compared to how nice reordering columns is 
- windows in a tab group have the same width, with the proposed "pinned/sticky" windows you could have windows with multiple widths
- the normal closing shorcut does not close the whole column (I dont know if there is a way to select the whole group as in i3/sway)
- touchpad gestures just dont work with tabs
- niri's tab groups are difficult to "discover", its hard to see what is inside or not without switching to the tab
- my mental model is "move horizontally for current workspace", "move vertically to switch workspaces". Tabs add an extra dimension and adds a bit more cognitive load to navigate

This is just purely personal and I understand that each of us could have different preferences :)

**Reply by cehteh** [reply]
I could even imagine to generalize this by being able to split a workspace vertically into 2 halves (or maybe 3 parts but no more)

Then each halve could be a independenty scolling entity. Or alternatively you have 2 workspaces on one monitor (sans status bar render and such).

Once this generalization is complete one can implement more functionality/constraints on top of that. Like opening new windows will open on the current colum instead right of it (mentioned in other tickets), configuring other layouts for such areas and so on.
Effectively making such an area 'pinned' by opening a new window will always be on the existing column and this column takes 100% of its split, then nothing scrolls and it appears static.

**Reply by nickjj** [reply]
After seeing your screenshots I think you can sort of hack around this use case with a "dummy" window and making your code editor a floating window.

- Open window 0 (a dummy app of your choice, all it does is take up space)
- Open window 1 (code editor)
- Open window 2 (browser)
- Open window 3 (docs)

At this point you have 4 windows and you scroll like normal.

- Make window 1 floating
- Position window 1 directly above window 0
- Start scrolling to the right and left as needed

The effect here is window 1 (code editor) will appear pinned on the left.

Nothing important ever ends up behind the floating window because window 0 (dummy app) blocks it. Your dummy app also never moves.

The main drawback with this approach is you can't use your normal key binds to scroll the non-pinned windows if window 1 is currently in focus (code editor) but how would you do this with a real non-hacky pinned approach? Is the idea that you'd scroll normally but the pinned window would never move, but it could still get focused like any other window?

To get around this drawback for now you can quickly tap alt-tab which will focus the last focused window instantly. This lets you jump from your fake-pinned window to any of the other windows and now you can scroll.

Not saying this is perfect or a replacement for pinned windows but maybe it's better than nothing for now?

**Comment by soymadip** [comment]
hello, is this going to be implemented?

i came across nsticky application, it works quite good but needs an extra process to be spawned.
it will make more sense to be built in to niri
